_overviews/scala3-book/scala-for-javascript-devs.md

```diff
@@ -0,0 +1,1238 @@
+enum Color:
+  case Red, Green, Blue
+```
+
+You can create a parameterized enum:  
+
+```scala
+enum Color(val rgb: Int):
+  case Red   extends Color(0xFF0000)
+  case Green extends Color(0x00FF00)
+  case Blue  extends Color(0x0000FF)
+```
+
+You can also create user-defined enum members:
+
+```scala
+enum Planet(mass: Double, radius: Double):
+  case Mercury extends Planet(3.303e+23, 2.4397e6)
+  case Venus   extends Planet(4.869e+24,6.0518e6)
+  case Earth   extends Planet(5.976e+24,6.37814e6)
+  // more planets here ...
+
+  private final val G = 6.67300E-11
+  def surfaceGravity = G * mass / (radius * radius)
+  def surfaceWeight(otherMass: Double) = otherMass * surfaceGravity
+```
+
+
+
+## Scala.js DOM Code
+
+Scala.js is a project that lets you write Scala code that is compiled to JavaScript code that can then be used in the browser. The approach is similar to CoffeeScript, TypeScript, and other languages that are compiled (“transpiled”) to JavaScript.  
+
+Once you include the necessary libraries, and import the necessary packages in your project, writing Scala.js code looks very similar to writing JavaScript code:  
+
+```scala
+// show an alert dialog on a button click
+$("#hello-button").click{() =>
+  dom.window.alert("Hello, world")
+}
+
+// define a button and what should happen when it’s clicked
+val btn = button(
+  "Click me",
+  onclick := { () =>
+    dom.window.alert("Hello, world")
+})
+
+// create two divs with css classes, an h2 element, and the button
+val content =
+  div(cls := "foo",
+    div(cls := "bar",
+      h2("Hello"),
+        btn
+      )
+  )
+
+// add the content to the DOM
+val root = dom.document.getElementById("root")
+root.innerHTML = ""
+root.appendChild(content.render)
+```
+
+Notice that although Scala is a type-safe language, no types are declared in that code. Scala’s strong type inference capabilites often make Scala code look like it’s dynamically-typed. But it is type-safe, so you catch many classes of errors early in the development cycle.  
+
+
+
+## Other Scala.js resources
+
+The Scala.js website has an excellent collection of tutorials from JavaScript developers interested in using Scala.js. Here are some of their initial tutorials:  
+
+- [Basic tutorial (creating a first Scala.js project)](https://www.scala-js.org/doc/tutorial/basic/)  
+- [Scala.js for JavaScript developers](https://www.scala-js.org/doc/sjs-for-js/)
+- [From ES6 to Scala: Basics  
+    ](https://www.scala-js.org/doc/sjs-for-js/es6-to-scala-part1.html)
+- [From ES6 to Scala: Collections](https://www.scala-js.org/doc/sjs-for-js/es6-to-scala-part2.html)
+- [From ES6 to Scala: Advanced](https://www.scala-js.org/doc/sjs-for-js/es6-to-scala-part3.html)  
+
+
+
+## Concepts that are unique to Scala 
+
+There are other concepts in Scala which currently have no equivalent in JavaScript:  
+
+- Almost everything related to contextual abstractions
+- Method features:
+  -   Multiple parameter lists
+  -   Default parameter values
```

Comment:
This also exists in JavaScript.

Reply:
Just to get you right, you are _only_ talking about default params, right?